src/strategies/latency.js

```diff
@@ -0,0 +1,231 @@
+/*
+ * moleculer
+ * Copyright (c) 2018 MoleculerJS (https://github.com/moleculerjs/moleculer)
+ * MIT Licensed
+ */
+"use strict";
+
+const _ = require("lodash");
+
+const { random } = require("lodash");
+const BaseStrategy = require("./base");
+
+/**
+ * Lowest latency invocation strategy
+ *
+ * Since Strategy can be instantiated multiple times, therefore,
+ * we need to have a "master" instance to send ping, and each
+ * individual "slave" instance will update their list dynamically
+ *
+ * These options can be configured in broker registry options:
+ *
+ * const broker = new ServiceBroker({
+ * 	logger: true,
+ * 	registry: {
+ * 		strategy: "LatencyStrategy",
+ * 		strategyOptions: {
+ * 			sampleCount: 5,
+ * 			lowLatency: 10,
+ * 			collectCount: 5,
+ * 			pingInterval: 10
+ * 		}
+ * 	}
+ * });
+ *
+ * @class LatencyStrategy
+ */
+class LatencyStrategy extends BaseStrategy {
+
+	constructor(registry, broker) {
+		super(registry, broker);
+
+		this.opts = _.defaultsDeep(registry.opts.strategyOptions, {
+			sampleCount: 5,
+			lowLatency: 10,
+			collectCount: 5,
+			pingInterval: 10
+		});
+
+		this.brokerStopped = false;
+
+		this.hostLatency = new Map();
+
+		/* hostMap contains:
+			hostname => {
+				historicLatency: [],
+				nodeList: []
+			}
+		*/
+		this.hostMap = new Map();
+
+		// short circuit
+		if (!this.broker.transit) return;
+
+		if (this.broker.localBus.listenerCount("$node.latencyMaster") === 0) {
+			// claim as master
+			this.broker.localBus.on("$node.latencyMaster", function() {});
+			// respond to PONG
+			this.broker.localBus.on("$node.pong", this.processPong.bind(this));
+			// dynamically add new node
+			this.broker.localBus.on("$node.connected", this.addNode.bind(this));
+			// dynamically remove node
+			this.broker.localBus.on("$node.disconnected", this.removeHostMap.bind(this));
+			// try to discovery all nodes on start up
+			this.broker.localBus.on("$broker.started", this.discovery.bind(this));
```

Comment:
It is not the best solution, because it is possible that the service starts after the broker started. In this case the timer won't start.
Would be better to call `this.discovery` directly here.

Reply:
That's actually okay. If `this.discovery` wasn't called, the module would just behave like RandomStrategy. The "discovery" doesn't have to happen on-time.

Reply:
\<sidenote>
`setInterval` might potentially cause problems (https://develoger.com/settimeout-vs-setinterval-cff85142555b). Usually, `setTimeout` is a safer invocation. 
\</sidenote>